### DISCUSSION TITLE
Improvement in ergonomics

### DISCUSSION THREAD

**Discussion by kishaningithub** [pkgxdev/pkgx#819] [Ideas]
This might sound like a rant (it is) but kindly take it as something which goes into the mind of a new comer who is trying to get started with pkgx

pkg has so many commands to remember (pkgx blah blah, dev blah blah, env blah blah). I just want to cd into a directory having file like .terraform-version, .python-version and all the installation and wireup should happen like magic. I should not even have to remember the name of this tool. Thats the best user experience in my opinion. Let me know if what i am asking is dumb.
I tried my best to use this tool and many other all in one tools(asdf, rtx etc) in the past but individual tools like tfenv, pyenv, nvm feel way more ergonomic than all the "all in one tools"

May be i am missing something which leads to this complexity in user experience. Kindly let me know.

**Comment by mxcl** [comment]
It's hard to write a tool that works for everyone. So it's understandable `pkgx` is maybe not the tool for you.

If you prefer to use all the different managers then you should do that (`pkgx` can provide those ofc but using `brew` or whatever may be your preference).

For your desired usecase you just have to type `dev`. It seems a little strange that you complain `pkgx` is too many commands, but instead you'd rather remember numerous other tools rather than one, but I probably am not understanding you properly.

Prior to version 1 the tool did actually work magically, you would step into a directory and it would just install everything and wire everything up. We called it “magic” and it felt pretty magical. It also was sometimes very surprising in a not good way. So we in the end, after reading numerous rants much like this one, we converted to this system.

You type `dev`. That’s it. I understand that one command is more than zero commands, and maybe we’ll allow the “magical” way to return with an option at some point, but we thought long and hard about the ergonomics of this tool and based on our knowledge of devs and the users we are targeting we believe our ergonomics match their expectations.

Thank you for your feedback.

**Reply by kishaningithub** [reply]
Thanks a lot for taking the time to respond, thanks to the above i now gave this tool a second try. Will add some context on what happened step by step which can help you look at the tool through my eyes.

### System details
Hardware: Macbook Pro with M1 chip
OS: Sonoma 14.0

## Installation journey
- I installed this using homebrew `brew install pkgxdev/made/pkgx`
- Generally after the installation the packages give out an info stating what to do next, this one did not give me that. Putting here a comparison between fzf (another brew package) and this one
```bash
# -- fzf tool
# Note the Caveats section here which exactly tells what i must do after.
# In this case it is asking me to run /opt/homebrew/opt/fzf/install
$ brew install fzf
... # Removing the downloading outputs for better clarity
==> Caveats
To install useful keybindings and fuzzy completion:
  /opt/homebrew/opt/fzf/install # Note this

To use fzf in Vim, add the following line to your .vimrc:
  set rtp+=/opt/homebrew/opt/fzf
==> Summary
🍺  /opt/homebrew/Cellar/fzf/0.43.0: 17 files, 3.9MB
...

# -- pkgx
# Here the caveats does mention running node@18 which is quite impressive
# but what i am looking for is shell bindings and cd hooks installation. Since i could not
# find it here i had to go to the website to find out how to do it. This is not a great
# UX for a terminal first developer like me in my opinion.
$ brew install pkgxdev/made/pkgx
...
==> Caveats
try it out:

    pkgx node@18 --eval 'console.log("pkgx: run anything")'
==> Summary
🍺  /opt/homebrew/Cellar/pkgx/1.0.2: 2 files, 103.2MB
```

- Added this line to my `~/.zshrc` file `source <(pkgx --shellcode)` based on the instruction from the website.
- Opened a new iterm tab then did the following on a terraform repo and a python repo

### My user journey - Terraform repo
- Went into a terraform repository having  a `.terraform-version` file
- Ran `dev`
```bash
$ dev
 × no env
```
- Then i thought may be i am doing something wrong or something is not available, lets try installing terraform using pkgx
```bash
$ pkgx install terraform
× multiple projects provide: terraform
│ pls be more specific:
│
│     pkgx install terraform.io
│     pkgx install github.com/tfutils/tfenv
│
╰─➤ https://docs.pkgx.sh/help/ambiguous-pkgspec
```
- Now im like, OK why would i need to even consider github.com/tfutils/tfenv given pkgx is gonna do the versioning for me and hence i did `pkgx install terraform.io` and this worked
- Then i ran `terraform version` which actually downloaded terraform and ran it. 
```
terraform version
 ✓  ~/.pkgx/terraform.io/v1.6.1
Terraform v1.6.1-dev
on darwin_arm64

Your version of Terraform is out of date! The latest version
is 1.6.1. You can update by downloading from https://www.terraform.io/downloads.html
```
- If you see above for some reason it was not respecting the terraform version in the `.terraform-version` file which is `0.13.7` can you kindly help on this one?

### My user journey - Python repo
- This journey was very smooth actually. In this repo there was a `.python-version` file having value of `3.9`
- When i went to the repo directory and did dev it crashed the first few times (unfortunately i lost that log :-( ) and after that it worked like magic
- The main problem here was did not know how to turn off the dev environment manually, so i tried running --help
```bash
$ dev --help
env +python.org~3.9
``` 
- After reading the website (as a terminal user i prefer never to see the website) i figured i have to run the `dev off` command and here's what happened
```bash
$ dev off
cmd not found: _pkgx_dev_off
```
- Then my impression went down on this and thought may be i should wait for a couple of releases before adopting this mainstream.

## Others
- I also played with `env +deno@1.33` I actually find this more scary than helpful. Can this be an "opt-in" feature?

**Reply by mxcl** [reply]
We haven't added support for `.terraform-version` yet. Like our docs say, if we are missing something then a ticket or pull request will add it.

I am surprised you are scared by the `env +deno` feature. Perhaps you don't understand what it does? per our docs, `env +deno` is simply applying the result of `pkgx +deno`. Type `pkgx +deno` and you can see the environment variables that get set. Per the docs `env +deno` is strictly temporary, the env is gone when you exit the Terminal session.

While I understand your desire that everything be listed in `--help`, it's impossible to do that. We have 40,000 words on docs.pkgx.sh and you just can't add that to `--help`.

`dev` is the least developed feature of `pkgx` currently, which is part of the reason for your confusing user journey. We are marketing pkgx as a tool for running pkgs with a bonus feature called `dev` that is built on top of the pkgx primitives. With time it will be complete. Currently it is not.

Thanks.

**Reply by kishaningithub** [reply]
Thanks a lot for the detailed answer that definitely reduced by fears on this one

Also just to clarify the --help IMO need not list every possible thing, that is definitely for the web documentation, it should give some examples of the most important ones may be. Some thing like https://github.com/tldr-pages/tldr

Given dev seems to be my most wanted tool, I will also start working on it and rise PRs for review 

Thanks for creating this! 😊

PS - I am a very big fan of homebrew, its literally the major reason why I use mac, have always even felt that apple should give a cut for homebrew for every macs they sell because many of my friends also use mac for that reason. May be tea protocol can somehow help with that?
